### PR TITLE
[Draft] LPS-194500 Make Configuration a partitioned control table

### PR DESCRIPTION
**Pull request by achaparro**
Hi @shuyangzhou,

I am sending this pull to you to verify if we are in the right direction to make the Configuration table as partitioned control table. This tasks is part of our epic to implement virtual instances migration in DB Partitioning environments.

Basically, now the Configuration table won't be a VIEW in non-default partitions and will be a table. For that, we need to be sure that:
1. We load the dictionaries for all partitions on startup. They should be able to inherit configurations from parent classes as before.
2. The CompanyThreadLocal is properly set when modifying Configuration via UI (this already worked)
3. The CompanyThreadLocal is properly set when loading config files via folder.
4. An upgrade converts the views in tables and migrate the scoped configurations to the proper partition.

This is still a draft because we need to implement tests yet.

Let me know what you think.

Thanks.

cc/ @LuisOrtiz89

### CODE REVIEW

**Comment by achaparro** [comment]
ci:test:sf

File: modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java [hunk 1]
```diff
@@ -265,6 +287,51 @@ StringPool.EQUAL, _escapeFilterValue(fileName),
...
+			(dictionary == null)) {
+
+			return 0L;
+		}
+
+		Long companyId = (Long)dictionary.get(scope.getPropertyKey());
+
+		if (companyId != null) {
+			if (!ArrayUtil.contains(
+					PortalInstancePool.getCompanyIds(), companyId)) {
```
Comment (shuyangzhou) [inline modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java:305]: @achaparro @LuisOrtiz89 configuration file processing logic starts way earlier than PortalInstancePool initialization, you may not see the existing company and have a false complain here. And eventually saves the configuration into the wrong db partition.

**Review by shuyangzhou — COMMENTED** [review]
I am also having a hard time to understand the logic in ConfigurationUpgradeProcess, please add some explanation.

File: modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java (hunk 1 above)
Comment (achaparro) [reply modules/apps/static/portal-file-install/portal-file-install-impl/src/main/java/com/liferay/portal/file/install/internal/configuration/ConfigurationFileInstaller.java:305]: got it @shuyangzhou, I will change it to use something reliable at this point

**Comment by achaparro** [comment]
Hey @shuyangzhou,

Regarding the upgrade process:
- Prior to this PR, Configuration was a control table and we store all the records in the default partition. Non-default partitions used a VIEW over that table.
- After this PR, all partitions will have their own Configuration table. We will store the scoped records in the proper partition 

So this is what the upgrade does:
1. Collect all scoped configurations that do not belong to the default company. They can be company, group or portlet-instance scope. For the two first cases we will remove them from the default company. SYSTEM configurations will remain in the default partition.
2- We iterate all partitions to copy the company or group configurations to the partition they belong. For portlet-instance configuration we will copy them to every partition since they can potentially apply to all partitions.
3- At the end, when we have iterated all partitions, we show a warning for those scope configurations that do not apply to any partition. They are just orphaned scope configurations.

During the process we also replace the Configuration views by the tables in the non-default partitions.

Please, take into account that the upgrade processes run once per company when DB Partitioning is enabled:
https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/upgrade/UpgradeProcess.java#L106

Let me know if now the process is clearer.

Thanks.

**Comment by shuyangzhou** [comment]
Ok, thanks for the explanation @achaparro ! Please fix the PortalInstances issue, then we can do another round of review.

**Pull request closed without merging**